### PR TITLE
provide Function._sage_; remove superfluous _sage_ methods; add more of them; fixes #3444

### LINKED ISSUE #3444: implement _sage_ methods
bc.. As agreed in this thread: http://groups.google.com/group/sympy/browse_thread/thread/34f0e3a7e598e13/dacdd250422a3e9e we need to implement _sage_ methods, so that sympy can be used in SAGE more
natively. This is high priority, because this will help sympy to be much
more widespread. And we need more developers.

p. Original issue for "#3444":https://github.com/sympy/sympy/issues/3444: "http://code.google.com/p/sympy/issues/detail?id=345":http://code.google.com/p/sympy/issues/detail?id=345

p. Original author: "https://code.google.com/u/104039945248245758823/":https://code.google.com/u/104039945248245758823/

### PR DESCRIPTION
**Pull request by rwst**
Adapted and squashed all commits of #8592 into one.

### CODE REVIEW

**Comment by skirpichev** [comment]
> Adapted and squashed all commits of #8592 into one.

Why you can't do this in #8592?  Please don't create new pr for this next time.

File: sympy/external/tests/test_sage.py [hunk 1]
```diff
@@ -114,7 +125,8 @@ def test_oo():
     assert sage.oo == sage.SR(sympy.oo)
     assert sympy.sympify(-sage.oo) == -sympy.oo
     assert -sage.oo == sage.SR(-sympy.oo)
-
+    #assert sympy.sympify(sage.UnsignedInfinityRing.gen()) == sympy.zoo
```
Comment (skirpichev) [inline sympy/external/tests/test_sage.py]: is this some todo?

File: sympy/core/function.py [hunk 2]
```diff
@@ -673,6 +673,13 @@ def _eval_as_leading_term(self, x):
         else:
             return self.func(*args)
 
+    def _sage_(self):
+        import sage.all as sage
+        fname = self.func.__name__
+        func = getattr(sage, fname)
```
Comment (skirpichev) [inline sympy/core/function.py]: Perhaps, you can add some private attribute to keep the "right" sage name?  This is an alternative for overriding _sage_ methods below.

Comment (skirpichev) [inline sympy/core/function.py]: But we can improve this latter too.

PR looks good for me.

File: sympy/external/tests/test_sage.py (hunk 1 above)
Comment (rwst) [inline sympy/external/tests/test_sage.py]: Yes, todo in Sage.

**Pull request merged**